README.md
InternetMessage.js
==================
[![NPM version][npm-badge]](http://badge.fury.io/js/internet-message)
[npm-badge]: https://badge.fury.io/js/internet-message.png

**InternetMessage.js** is a small JavaScript library for parsing messages and
stringifying objects to the syntax of [RFC 733 (ARPA Network Text
Message)][rfc733], [RFC 822 (ARPA Internet Text Messages)][rfc822] and [RFC 2822
(Internet Message Format)][rfc2822]. You've probably seen the format in e-mail
messages or in HTTP. It's basically a **format for headers and a body**. You can
use it to **send both text or binary data**.

InternetMessage.js isn't meant to be an e-mail or HTTP parser, but more of
a useful small library for sending **standard formatted messages** over any
channel that doesn't have built-in structured data. Comes in handy with message
queues (like [ZeroMQ][zeromq]), event sockets ([WebSockets][websockets] and
[Server-Sent Events][sse]) and such. Where the medium just gives you a single
blob without structure, use InternetMessage.js and [RFC 822][rfc822] to not have
to invent a custom format.

#### Example
```
Message-Id: fc00fc02a215412780bf09a7dcd5e33c
Content-Type: application/json

{"type":"created", "uri": "/models/1"}
```

[rfc733]: https://tools.ietf.org/html/rfc733
[rfc822]: https://tools.ietf.org/html/rfc822
[rfc2822]: https://tools.ietf.org/html/rfc2822
[zeromq]: http://zeromq.org/
[websockets]: https://tools.ietf.org/html/rfc6455
[sse]: http://dev.w3.org/html5/eventsource/


Installing
----------
```sh
npm install internet-message
```

InternetMessage.js follows [semantic versioning](http://semver.org/), so feel
free to depend on its major version with something like `>= 1.0.0 < 2`
(a.k.a `^1.0.0`).


Using
-----
Create an instance of `InternetMessage` by giving it an object of headers and
any text for the body.
```javascript
var InternetMessage = require("internet-message")

var msg = new InternetMessage({
  "Content-Type": "application/json",
  "Location": "http://example.com/models/1"
}, JSON.stringify({name: "John"}))
```

Calling `msg.toString()` will then return the message as a string:
```
Content-Type: application/json
Location: http://example.com/models/1

{"name":"John"}
```

As the standard requires, lines will end with CRLF (carriage return and line
feed).

You can also use `InternetMessage.stringify` directly without creating an
intermediate `InternetMessage` instance:
```javascript
InternetMessage.stringify(headers, body)
```

### Parsing
Giving the message below to `InternetMessage.parse` will give you an instance of
`InternetMessage`:
```
Content-Type: application/json
Location: http://example.com/models/1

{"name":"John"}
```

The message will have headers as enumerable properties and a `body` property
with the body, if it has one. All header names are in lower-case for easier
access.

```javascript
var msg = InternetMessage.parse(TEXT)
msg["content-type"] // => "application/json"
msg["location"]     // => "http://example.com/models/1"
msg.body            // => "{\"name\":\"John\"}"
```

### Customizing end-of-line or start-of-body
If you wish to customize the end-of-line and start-of-body characters the header
uses, pass them as strings of any length to `InternetMessage.prototype.toString`
or `InternetMessage.stringify`.
```javascript
msg.toString({eol: "\n"})
msg.toString({eol: "\x1e", sob: "\x02"})
InternetMessage.stringify(msg, {eol: "\n"})
```

Remember to pass the same options later to `InternetMessage.parse`.

If you've just changed the `eol` option to `\n`, then don't bother.
`InternetMessage.parse` supports both `\r\n` and `\n` as the end-of-line out of
the box.


API
---
For extended documentation on all functions, please see the [InternetMessage.js
API Documentation][api].

[api]: https://github.com/moll/js-internet-message/blob/master/doc/API.md

### [InternetMessage](https://github.com/moll/js-internet-message/blob/master/doc/API.md#InternetMessage)
- [_header_](https://github.com/moll/js-internet-message/blob/master/doc/API.md#internetMessage._header_)
- [body](https://github.com/moll/js-internet-message/blob/master/doc/API.md#internetMessage.body)
- [parse](https://github.com/moll/js-internet-message/blob/master/doc/API.md#InternetMessage.parse)(message, [options])
- [stringify](https://github.com/moll/js-internet-message/blob/master/doc/API.md#InternetMessage.stringify)(message, [body], [options])
- [toJSON](https://github.com/moll/js-internet-message/blob/master/doc/API.md#InternetMessage.prototype.toJSON)()
- [toString](https://github.com/moll/js-internet-message/blob/master/doc/API.md#InternetMessage.prototype.toString)([options])


License
-------
InternetMessage.js is released under a *Lesser GNU Affero General Public
License*, which in summary means:

- You **can** use this program for **no cost**.
- You **can** use this program for **both personal and commercial reasons**.
- You **do not have to share your own program's code** which uses this program.
- You **have to share modifications** (e.g. bug-fixes) you've made to this
  program.

For more convoluted language, see the `LICENSE` file.


About
-----
**[Andri Möll][moll]** typed this and the code.  
[Monday Calendar][monday] supported the engineering work.

If you find InternetMessage.js needs improving, please don't hesitate to type to
me now at [andri@dot.ee][email] or [create an issue online][issues].

[email]: mailto:andri@dot.ee
[issues]: https://github.com/moll/js-internet-message/issues
[moll]: http://themoll.com
[monday]: https://mondayapp.com


CHANGELOG.md
## 1.0.0 (Jan 18, 2015)
- Good old days.


test/index_test.js
var InternetMessage = require("..")

describe("InternetMessage", function() {
  describe("new", function() {
    it("must be an instance of InternetMessage", function() {
      new InternetMessage().must.be.an.instanceof(InternetMessage)
    })

    it("must set headers and body if given separately", function() {
      var headers = {"Content-Type": "text/plain", "Range": "100-200"}
      var msg = new InternetMessage(headers, "Hello")
      msg.must.have.property("content-type", "text/plain")
      msg.must.have.property("range", "100-200")
      msg.must.have.property("body", "Hello")
    })

    it("must set headers and body if given in one object", function() {
      var obj = {"Content-Type": "text/plain", "Range": "100-200", body: "Hi"}
      var msg = new InternetMessage(obj)
      msg.must.have.property("content-type", "text/plain")
      msg.must.have.property("range", "100-200")
      msg.must.have.property("body", "Hi")
    })

    it("must not set body if given undefined", function() {
      new InternetMessage(null, undefined).must.not.have.property("body")
    })

    it("must set body if given null", function() {
      new InternetMessage(null, null).must.have.property("body", null)
    })

    it("must convert header names to lower case", function() {
      var msg = new InternetMessage({"CoNTenT-TyPe": "text/plain"})
      msg.must.have.property("content-type", "text/plain")
    })

    it("must not have other keys besides header and body", function() {
      var msg = new InternetMessage({"Content-Type": "text/plain"}, "Hello")
      msg.must.have.keys(["content-type", "body"])
    })

    it("must set headers and body given InternetMessage", function() {
      var other = new InternetMessage({"Content-Type": "text/plain"}, "Hello")
      var msg = new InternetMessage(other)
      msg.must.eql(other)
    })

    it("must not lose header casing if given InternetMessage", function() {
      var other = new InternetMessage({"Content-Type": "text/plain"}, "Hello")
      var msg = new InternetMessage(other)
      msg.toJSON().must.eql({"Content-Type": "text/plain", body: "Hello"})
    })
  })

  describe(".prototype.toString", function() {
    it("must stringify header and body", function() {
      var headers = {"Content-Type": "text/plain"}
      var msg = new InternetMessage(headers, "Hello")
      msg.toString().must.equal("Content-Type: text/plain\r\n\r\nHello")
    })

    it("must stringify header and body given options", function() {
      var headers = {"Content-Type": "text/plain"}
      var msg = new InternetMessage(headers, "Hello")
      var str = msg.toString({eol: "\x1e", sob: "\x02"})
      str.must.equal("Content-Type: text/plain\x1e\x02Hello")
    })

    it("must stringify header in given capitalization", function() {
      var headers = {"CoNTenT-TyPe": "text/plain"}
      var msg = new InternetMessage(headers, "Hello")
      msg.toString().must.equal("CoNTenT-TyPe: text/plain\r\n\r\nHello")
    })
  })

  describe(".prototype.toJSON", function() {
    it("must return an object with header and body", function() {
      var msg = new InternetMessage({"Content-Type": "text/plain"}, "Hello")
      msg.toJSON().must.eql({"Content-Type": "text/plain", body: "Hello"})
    })
  })

  describe(".parse", function() {
    it("must parse header and body", function() {
      var msg = InternetMessage.parse("Content-Type: text/plain\r\n\r\nHello")
      msg.must.eql(new InternetMessage({"Content-Type": "text/plain"}, "Hello"))
    })

    it("must parse header and body given only line feed", function() {
      var msg = InternetMessage.parse("Content-Type: text/plain\n\nHello")
      msg.must.eql(new InternetMessage({"Content-Type": "text/plain"}, "Hello"))
    })

    it("must parse header and body given end of line", function() {
      var text = "Content-Type: text/plain\x1e\x1eHello"
      var msg = InternetMessage.parse(text, {eol: "\x1e"})
      msg.must.eql(new InternetMessage({"Content-Type": "text/plain"}, "Hello"))
    })

    it("must parse header and body given longer end of line", function() {
      var text = "Content-Type: text/plain######Hello"
      var msg = InternetMessage.parse(text, {eol: "###"})
      msg.must.eql(new InternetMessage({"Content-Type": "text/plain"}, "Hello"))
    })

    it("must parse header and body given start of body", function() {
      var text = "Content-Type: text/plain\r\n\x02Hello"
      var msg = InternetMessage.parse(text, {sob: "\x02"})
      msg.must.eql(new InternetMessage({"Content-Type": "text/plain"}, "Hello"))
    })

    it("must parse header and body given longer start of body", function() {
      var text = "Content-Type: text/plain\r\n###Hello"
      var msg = InternetMessage.parse(text, {sob: "###"})
      msg.must.eql(new InternetMessage({"Content-Type": "text/plain"}, "Hello"))
    })

    it("must parse headers and body", function() {
      var str = "Content-Type: text/plain\r\nRange: 100-200\r\n\r\nHello"
      var msg = InternetMessage.parse(str)
      var headers = {"Content-Type": "text/plain", "Range": "100-200"}
      msg.must.eql(new InternetMessage(headers, "Hello"))
    })

    it("must parse headers and body given end of line", function() {
      var str = "Content-Type: text/plain\x1eRange: 100-200\x1e\x1eHello"
      var msg = InternetMessage.parse(str, {eol: "\x1e"})
      var headers = {"Content-Type": "text/plain", "Range": "100-200"}
      msg.must.eql(new InternetMessage(headers, "Hello"))
    })

    it("must parse headers and body given start of body", function() {
      var str = "Content-Type: text/plain\r\nRange: 100-200\r\n\x02Hello"
      var msg = InternetMessage.parse(str, {sob: "\x02"})
      var headers = {"Content-Type": "text/plain", "Range": "100-200"}
      msg.must.eql(new InternetMessage(headers, "Hello"))
    })

    it("must parse header without leading space", function() {
      var msg = InternetMessage.parse("Content-Type:text/plain\r\n")
      msg.must.eql(new InternetMessage({"Content-Type": "text/plain"}))
    })

    it("must parse header with multiple colons", function() {
      var msg = InternetMessage.parse("Ratio: 1:2\r\n")
      msg.must.eql(new InternetMessage({"Ratio": "1:2"}))
    })

    it("must parse headers with no body", function() {
      var msg = InternetMessage.parse("Content-Type: text/plain\r\n")
      msg.must.eql(new InternetMessage({"Content-Type": "text/plain"}))
    })

    it("must parse headers with empty body", function() {
      var msg = InternetMessage.parse("Content-Type: text/plain\r\n\r\n")
      msg.must.eql(new InternetMessage({"Content-Type": "text/plain"}, ""))
    })

    it("must parse headers with blank lined body", function() {
      var msg = InternetMessage.parse("Content-Type: text/plain\r\n\r\n\r\n")
      msg.must.eql(new InternetMessage({"Content-Type": "text/plain"}, "\r\n"))
    })

    // An empty string has to be valid as the body is optional and there's no
    // requirement to have any headers either.
    it("must parse given empty string", function() {
      var msg = InternetMessage.parse("")
      msg.must.eql(new InternetMessage)
    })

    it("must parse body given just body", function() {
      var msg = InternetMessage.parse("\r\nHello")
      msg.must.eql(new InternetMessage(null, "Hello"))
    })

    it("must parse body given just body and line feed", function() {
      var msg = InternetMessage.parse("\nHello")
      msg.must.eql(new InternetMessage(null, "Hello"))
    })

    it("must parse body given just blank body", function() {
      var msg = InternetMessage.parse("\r\n")
      msg.must.eql(new InternetMessage(null, ""))
    })

    it("must parse body given just blank body and line feed", function() {
      var msg = InternetMessage.parse("\n")
      msg.must.eql(new InternetMessage(null, ""))
    })

    it("must throw SyntaxError given a line with no end of line", function() {
      var err
      try { InternetMessage.parse("Content-Type: text/plain") }
      catch (ex) { err = ex }
      err.must.be.an.instanceof(SyntaxError)
    })

    it("must throw SyntaxError given a blank line", function() {
      var err
      try { InternetMessage.parse(" ") } catch (ex) { err = ex }
      err.must.be.an.instanceof(SyntaxError)
    })
  })

  describe(".stringify", function() {
    it("must stringify InternetMessage", function() {
      var headers = {"Content-Type": "text/plain", "Range": "100-200"}
      var msg = new InternetMessage(headers, "Hello")
      var str = InternetMessage.stringify(msg)
      str.must.equal("Content-Type: text/plain\r\nRange: 100-200\r\n\r\nHello")
    })

    it("must stringify InternetMessage given options", function() {
      var headers = {"Content-Type": "text/plain", "Range": "100-200"}
      var msg = new InternetMessage(headers, "Hello")
      var str = InternetMessage.stringify(msg, {eol: "\x1e", sob: "\x02"})
      str.must.equal("Content-Type: text/plain\x1eRange: 100-200\x1e\x02Hello")
    })

    it("must stringify header and body", function() {
      var headers = {"Content-Type": "text/plain"}
      var str = InternetMessage.stringify(headers, "Hello")
      str.must.equal("Content-Type: text/plain\r\n\r\nHello")
    })

    it("must stringify header and body given end of line", function() {
      var headers = {"Content-Type": "text/plain"}
      var str = InternetMessage.stringify(headers, "Hello", {eol: "\x1e"})
      str.must.equal("Content-Type: text/plain\x1e\x1eHello")
    })

    it("must stringify header and body given start of body", function() {
      var headers = {"Content-Type": "text/plain"}
      var str = InternetMessage.stringify(headers, "Hello", {sob: "\x02"})
      str.must.equal("Content-Type: text/plain\r\n\x02Hello")
    })

    it("must stringify headers and body", function() {
      var headers = {"Content-Type": "text/plain", "Range": "100-200"}
      var str = InternetMessage.stringify(headers, "Hello")
      str.must.equal("Content-Type: text/plain\r\nRange: 100-200\r\n\r\nHello")
    })

    it("must stringify headers and body given end of line", function() {
      var headers = {"Content-Type": "text/plain", "Range": "100-200"}
      var str = InternetMessage.stringify(headers, "Hello", {eol: "\x1e"})
      str.must.equal("Content-Type: text/plain\x1eRange: 100-200\x1e\x1eHello")
    })

    it("must stringify headers and body given start of body", function() {
      var headers = {"Content-Type": "text/plain", "Range": "100-200"}
      var str = InternetMessage.stringify(headers, "Hello", {sob: "\x02"})
      str.must.equal("Content-Type: text/plain\r\nRange: 100-200\r\n\x02Hello")
    })

    it("must stringify body given empty headers", function() {
      InternetMessage.stringify({}, "Hello").must.equal("\r\nHello")
    })

    it("must stringify headers and empty body if given undefined", function() {
      var headers = {"Content-Type": "text/plain", "Range": "100-200"}
      var str = InternetMessage.stringify(headers, undefined)
      str.must.equal("Content-Type: text/plain\r\nRange: 100-200\r\n")
    })

    it("must stringify headers and empty body if given null", function() {
      var headers = {"Content-Type": "text/plain", "Range": "100-200"}
      var str = InternetMessage.stringify(headers, null)
      str.must.equal("Content-Type: text/plain\r\nRange: 100-200\r\n")
    })

    it("must stringify body given undefined headers", function() {
      InternetMessage.stringify(undefined, "Hello").must.equal("\r\nHello")
    })

    it("must stringify body given null headers", function() {
      InternetMessage.stringify(null, "Hello").must.equal("\r\nHello")
    })

    it("must stringify empty if both undefined", function() {
      InternetMessage.stringify(undefined, undefined).must.equal("")
    })

    it("must stringify empty if both null", function() {
      InternetMessage.stringify(null, null).must.equal("")
    })
  })
})


index.js
var LF = "\n"
var CR = "\r"
var CRLF = CR + LF
module.exports = InternetMessage

/**
 * Create an instance of `InternetMessage`.
 *
 * The only enumerable properties of an internet message will be the headers
 * you give to it and an optional `body`. Feel free to use a `for-in` loop.
 *
 * For a message with no body, don't pass one or set it to `null` or
 * `undefined` as the `body` argument.
 *
 * [rfc733]: https://tools.ietf.org/html/rfc733
 * [rfc822]: https://tools.ietf.org/html/rfc822
 * [rfc2822]: https://tools.ietf.org/html/rfc2822
 * [`InternetMessage`]: #InternetMessage
 * [`InternetMessage.stringify`]: #InternetMessage.stringify
 *
 * @class InternetMessage
 * @constructor
 * @param {Object} headers
 * @param {String} [body]
 */
function InternetMessage(msg, body) {
  if (msg instanceof InternetMessage) msg = msg.toJSON()

  var name, names = {}
  for (name in msg) names[name.toLowerCase()] = name
  define(this, "_headers", names)

  for (name in names) this[name] = msg[names[name]]
  if (body !== undefined) this.body = body
}

/**
 * Set to the header value.  
 * Headers are always lower-cased, but their original capitalization is
 * retained internally for later stringifying.
 *
 * @property {String} _header_
 */

/**
 * Set to the body of the message.
 *
 * If the message had no body (didn't contain any start-of-body character),
 * it'll be left out entirely. Accessing it will return the default
 * JavaScript's `undefined`.
 *
 * If there was an empty body, e.g for the message below, `body` will be set to
 * `""`.
 * ```
 * Content-Type: text/plain
 *
 * ```
 *
 * @property {String} body
 */

/**
 * Calls [`InternetMessage.stringify`][] with itself, forwarding any options
 * given to it.
 *
 * @example
 * var msg = new InternetMessage({"Content-Type": "text/plain"}, "Hello")
 * msg.toString({eol: "\n")
 *
 * @method toString
 * @param {Object} [options]
 */
define(InternetMessage.prototype, "toString", function(opts) {
  return this.constructor.stringify(this, opts)
})

/**
 * Returns a plain object of headers and `body`.  
 * Headers will be in their original capitalization.
 *
 * @method toJSON
 */
define(InternetMessage.prototype, "toJSON", function() {
  var obj = {}

  for (var name in this) {
    var value = this[name]
    if (name in this._headers) name = this._headers[name]
    obj[name] = value
  }

  return obj
})

/**
 * Parse a string following [RFC 733][rfc733], [RFC 822][rfc822] or
 * [RFC 2822][rfc2822] to an instance of [`InternetMessage`][].  
 * One exception though: It doesn't yet support wrapped or folded header
 * lines.  
 * Throws `SyntaxError` if there's an invalid header line.
 *
 * An empty message (`""`) is valid and means it had no headers nor a body.
 * Both are optional.  
 * A message consiting of whitespace (`" "`) however isn't valid and results in
 * a `SyntaxError`.
 *
 * **Options**:
 *
 * Name | Value
 * -----|------
 * eol  | String at the end of header lines.<br>Defaults to `\n` with an optional preceding `\r`.
 * sob  | String at the end of all headers to signal the start of body.<br>If not given, same as `eol`.
 *
 * @static
 * @method parse
 * @param {String} message
 * @param {Object} [options]
 */
InternetMessage.parse = function(msg, opts) {
  var eol = opts != null && "eol" in opts ? opts.eol : LF
  var sob = opts != null && "sob" in opts ? opts.sob : eol
  var body = opts != null && "body" in opts ? !!opts.body : true

  var headers = []
  for (var i = 0; i < msg.length; i = eolIndex + eol.length) {
    if (stringAt(msg, sob, i)) break
    if (sob == LF && stringAt(msg, CRLF, i)) { ++i; break }

    var eolIndex = msg.indexOf(eol, i)
    if (eolIndex == -1) throw new SyntaxError("Invalid Message: No EOL")

    var line = msg.slice(i, eolIndex)
    if (eol == LF && line[line.length - 1] == CR) line = line.slice(0, -1)
    headers.push(line)
  }

  return new InternetMessage(
    parseHeaders(headers),
    body && i < msg.length ? msg.slice(i + sob.length) : undefined
  )
}

/**
 * Returns a string following [RFC 733][rfc733], [RFC 822][rfc822] and
 * [RFC 2822][rfc2822].  
 * One exception though: It doesn't wrap long header lines as suggested by
 * those RFCs.
 *
 * Pass it an instance of [`InternetMessage`][] or a plain object of headers
 * and a body.
 *
 * **Options**:
 *
 * Name | Value
 * -----|------
 * eol  | String to use at the end of header lines.<br>Defaults to `\r\n`.
 * sob  | String to use at the end of all headers to signal the start of body.<br>If not given, same as `eol`.
 *
 * @example
 * var msg = new InternetMessage({"Content-Type": "text/plain"}, "Hello")
 * var opts = {eol: "\n"}
 *
 * InternetMessage.stringify(msg, opts)
 * InternetMessage.stringify({"Content-Type": "text/plain"}, "Hello", opts)
 *
 * @static
 * @method stringify
 * @param {Object} message
 * @param {String} [body]
 * @param {Object} [options]
 */
InternetMessage.stringify = function(msg, body, opts) {
  if (msg instanceof InternetMessage)
    msg = msg.toJSON(), opts = body, body = msg.body
  else if (msg != null && body === undefined)
    body = msg.body

  var eol = opts != null && "eol" in opts ? opts.eol : CRLF
  var sob = opts != null && "sob" in opts ? opts.sob : eol

  var str = ""
  for (var name in msg) if (name != "body") str += name + ": " + msg[name] + eol
  if (body != null) str += sob + body

  return str
}

function define(obj, name, value) {
  Object.defineProperty(obj, name, {
    value: value, configurable: true, writable: true
  })
}

function parseHeaders(lines) {
  if (lines.length == 0) return null

  return lines.reduce(function(headers, header) {
    var pos = header.indexOf(":")
    var name = header.slice(0, pos)
    if (name != "") headers[name] = trimLeftSpace(header.slice(pos + 1))
    return headers
  }, {})
}

function stringAt(hay, str, i) { return hay.substr(i, str.length) == str }
function trimLeftSpace(str) { return str[0] == " " ? str.slice(1) : str }
